core/extraction/towns_fund.py

```diff
@@ -0,0 +1,454 @@
+"""
+Methods specifically for extracting data from Towns Fund reporting template (Excel Spreadsheet)
+"""
+import numpy as np
+import pandas as pd
+
+
+def ingest_towns_fund_data(df_ingest: pd.DataFrame) -> dict[pd.DataFrame]:
+    """
+    Extract data from Towns Fund Reporting Template into column headed Pandas DataFrames.
+
+    :param df_ingest: DataFrame of parsed Excel data.
+    :return: Dictionary of extracted "tables" as DataFrames.
+    """
+
+    # TODO: check data for all DM v3.7 tabs is now extracted
+    towns_fund_extracted = {"df_programme_extracted": extract_programme(df_ingest["2 - Project Admin"])}
+    towns_fund_extracted["df_projects_extracted"] = extract_project(df_ingest["2 - Project Admin"])
+    number_of_projects = len(towns_fund_extracted["df_projects_extracted"].index)
+    towns_fund_extracted["df_programme_progress_extracted"] = extract_programme_progress(
+        df_ingest["3 - Programme Progress"]
+    )
+    towns_fund_extracted["df_project_progress_extracted"] = extract_project_progress(
+        df_ingest["3 - Programme Progress"]
+    )
+    towns_fund_extracted["df_funding_questions_extracted"] = extract_funding_questions(
+        df_ingest["4a - Funding Profiles"]
+    )
+    towns_fund_extracted["df_funding_comments_extracted"] = extract_funding_comments(
+        df_ingest["4a - Funding Profiles"], number_of_projects
+    )
+    towns_fund_extracted["df_funding_extracted"] = extract_funding_data(
+        df_ingest["4a - Funding Profiles"], number_of_projects
+    )
+
+    towns_fund_extracted["df_psi_extracted"] = extract_psi(df_ingest["4b - PSI"])
+
+    towns_fund_extracted["df_outputs_extracted"] = extract_outputs(df_ingest["5 - Project Outputs"], number_of_projects)
+    towns_fund_extracted["df_outcomes_extracted"] = extract_outcomes(df_ingest["6 - Outcomes"])
```

Comment:
IDE is complaining saying these could all be put into one big dictionary instantiation but I guess that's not quite as easy to debug?..

Reply:
yeah I thought about thta, but decided to leave it verbosely line-by-line, for the purpose of debugging, as you say